### PR TITLE
GDPR-62 Checking and recording Pathfinder referral

### PR DESCRIPTION
**Pull request by brightonsbox**
The comments below have been addressed now and this PR re-opened.

Now, to make it easy to extend the list of retention reasons, the RETENTION_REASON table 
will just have a REASON_CODE field. Any additional data about that reason will be provided in a linked table.

### CODE REVIEW

File: src/main/java/uk/gov/justice/hmpps/datacompliance/repository/jpa/model/retention/RetentionReason.java [hunk 1]
```diff
@@ -42,4 +42,7 @@
     @ManyToOne
     @JoinColumn(name = "MANUAL_RETENTION_ID")
     private ManualRetention manualRetention;
+
```
Comment (andymarke) [inline src/main/java/uk/gov/justice/hmpps/datacompliance/repository/jpa/model/retention/RetentionReason.java:45]: Will there always be a small number of retention reasons or will there be an ever growing number of columns on this table?

File: src/main/java/uk/gov/justice/hmpps/datacompliance/services/retention/RetentionService.java [hunk 2]
```diff
@@ -11,19 +12,27 @@
...
-        return manualRetentionService.findManualOffenderRetentionWithReasons(offenderNumber)
-                .map(manualRetention ->
-                        RetentionReason.builder()
-                                .manualRetention(manualRetention)
-                                .build());
+        final var isReferredToPathfinder = pathfinderApiClient.isReferredToPathfinder(offenderNumber);
+        final var manualRetention = manualRetentionService.findManualOffenderRetentionWithReasons(offenderNumber);
+
+        if (!isReferredToPathfinder && manualRetention.isEmpty()) {
+            return Optional.empty();
```
Comment (andrewrlee) [inline src/main/java/uk/gov/justice/hmpps/datacompliance/services/retention/RetentionService.java:28]: worth logging which one it is?

File: src/main/java/uk/gov/justice/hmpps/datacompliance/services/retention/RetentionService.java [hunk 3]
```diff
@@ -11,19 +12,27 @@
...
+        }
+
+        final var retentionReason = RetentionReason.builder()
+                .pathfinderReferred(isReferredToPathfinder);
+
+        manualRetention.ifPresent(retentionReason::manualRetention);
+
+        return Optional.of(retentionReason.build());
```
Comment (andrewrlee) [inline src/main/java/uk/gov/justice/hmpps/datacompliance/services/retention/RetentionService.java:36]: worth logging any of these responses?

**Comment by brightonsbox** [comment]
As discussed with Andy M, I'm going to reword the table structure slightly to accommodate a potentially growing set of retention reasons, avoiding the need to keep adding new columns for every reason

File: src/main/java/uk/gov/justice/hmpps/datacompliance/repository/jpa/model/retention/RetentionReason.java (hunk 1 above)
Comment (brightonsbox) [reply src/main/java/uk/gov/justice/hmpps/datacompliance/repository/jpa/model/retention/RetentionReason.java]: Reworked

File: src/main/java/uk/gov/justice/hmpps/datacompliance/services/retention/RetentionService.java (hunk 3 above)
Comment (brightonsbox) [reply src/main/java/uk/gov/justice/hmpps/datacompliance/services/retention/RetentionService.java]: Logging added

File: src/main/java/uk/gov/justice/hmpps/datacompliance/repository/jpa/model/retention/RetentionReasonManual.java [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+@DiscriminatorValue(RetentionReasonManual.MANUAL_RETENTION)
+public class RetentionReasonManual extends RetentionReason {
+
+    public static final String MANUAL_RETENTION = "MANUAL_RETENTION";
+
+    @OneToOne(mappedBy = "retentionReason", cascade = CascadeType.PERSIST)
+    private ManualRetentionLinkTable manualRetentionLinkTable;
+
+    public RetentionReasonManual() {
+        super(null, null, MANUAL_RETENTION);
```
Comment (brightonsbox) [inline src/main/java/uk/gov/justice/hmpps/datacompliance/repository/jpa/model/retention/RetentionReasonManual.java:33]: Passing nulls for ids that we can't provide yet at point of creation

**Pull request merged**